drivers/at86rf2xx/aes_spi.c

```diff
@@ -0,0 +1,425 @@
+{
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+                  AT86RF2XX_AES_CTRL_AES_DIR__DEC;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, false, &cfg, NULL, sizeof(cfg));
+    _aes_open_read(dev, AT86RF2XX_REG__AES_KEY_START);
+    _aes_transfer_bytes(dev, false, NULL, key, AT86RF2XX_AES_KEY_LENGTH);
+    at86rf2xx_spi_release_bus(dev);
+}
+
+void at86rf2xx_aes_key_write_decrypt(at86rf2xx_t *dev,
+                                     const uint8_t key[AT86RF2XX_AES_KEY_LENGTH])
+{
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+                  AT86RF2XX_AES_CTRL_AES_DIR__DEC;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+    _aes_transfer_bytes(dev, false, key, NULL, AT86RF2XX_AES_KEY_LENGTH);
+    at86rf2xx_spi_release_bus(dev);
+}
+
+int at86rf2xx_aes_ecb_encrypt(at86rf2xx_t *dev,
+                              aes_block_ptr_t cipher,
+                              uint8_t key[AT86RF2XX_AES_BLOCK_SIZE],
+                              const aes_block_ptr_t plain,
+                              uint8_t nblocks)
+{
+    if (!nblocks) {
+        return 0;
+    }
+    uint8_t cfg = AT86RF2XX_AES_CTRL_AES_MODE__ECB |
+                  AT86RF2XX_AES_CTRL_AES_DIR__ENC;
+    uint8_t mirror = cfg | AT86RF2XX_AES_CTRL_MIRROR_MASK__AES_REQUEST;
+    at86rf2xx_spi_get_bus(dev);
+    _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+    _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+    _aes_transfer_bytes(dev, true, plain[0], NULL, AT86RF2XX_AES_BLOCK_SIZE);
+    _aes_transfer_bytes(dev, false, &mirror, NULL, sizeof(mirror));
+
+    if (_aes_wait_for_result(dev) & AT86RF2XX_AES_STATUS_MASK__AES_ER) {
+        at86rf2xx_spi_release_bus(dev);
+        return -EIO;
+    }
+    if (key) {
+        cfg = AT86RF2XX_AES_CTRL_AES_MODE__KEY |
+              AT86RF2XX_AES_CTRL_AES_DIR__DEC;
+        _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+        _aes_transfer_bytes(dev, false, &cfg, NULL, sizeof(cfg));
+        _aes_open_read(dev, AT86RF2XX_REG__AES_KEY_START);
+        _aes_transfer_bytes(dev, false, NULL, key, AT86RF2XX_AES_KEY_LENGTH);
+    }
+    cfg = AT86RF2XX_AES_CTRL_AES_MODE__ECB |
+          AT86RF2XX_AES_CTRL_AES_DIR__ENC;
+    for (unsigned i = 1; i < nblocks; i++) {
+        _aes_open_write(dev, AT86RF2XX_REG__AES_CTRL);
+        _aes_transfer_bytes(dev, true, &cfg, NULL, sizeof(cfg));
+        _aes_transfer_bytes(dev, true, plain[i], NULL, 1);
+        _aes_transfer_bytes(dev, true, plain[i] + 1,
+                            cipher ? cipher[i - 1] : NULL,
+                            AT86RF2XX_AES_BLOCK_SIZE - 1);
+        _aes_transfer_bytes(dev, false, &mirror,
+                            cipher ? cipher[i - 1] + AT86RF2XX_AES_BLOCK_SIZE - 1 : NULL,
+                            1);
```

Comment:
I see this block a lot - can it be moved to a helper function?

Reply:
yes I can do that.